sympy/matrices/dense.py

```diff
@@ -131,14 +132,18 @@ def __setitem__(self, key, value):
     def _cholesky(self):
         """Helper function of cholesky.
         Without the error checks.
-        To be used privately. """
+        To be used privately.
+        Implements the Cholesky–Banachiewicz algorithm.
+        """
         L = zeros(self.rows, self.rows)
         for i in range(self.rows):
             for j in range(i):
                 L[i, j] = (1 / L[j, j])*(self[i, j] -
-                                         sum(L[i, k]*L[j, k] for k in range(j)))
-            L[i, i] = sqrt(self[i, i] -
-                           sum(L[i, k]**2 for k in range(i)))
+                    sum(L[i, k]*L[j, k].conjugate() for k in range(j)))
+            Lii2 = self[i, i] - sum(Abs(L[i, k])**2 for k in range(i))
```

Comment:
It would probably be more efficient to write `L[i, k]*L[i, k].conjugate()`.

Reply:
That's what I wrote initially, and I'm sure it would be more efficient if SymPy treated complex numbers like it does real numbers (evaluating their products). But it does not:  
```
>>> Abs(2 + I)**2
5
>>> (2 + I)*(2 - I)
(2 - I)*(2 + I)
```
As a result, instead of a single positive number Lii2 becomes a long expression like
```
(2 - I)*(2 + I) + (3 - 4*I)*(3 + 4*I) + ... 
```
and this snowballs through the computation. 

`expand_mul(L[i, k]*L[i, k].conjugate())` could work, but in my timed tests it was slightly slower than `Abs(L[i, k])**2`

Reply:
On the other hand, perhaps using `expand_mul` throughout this computation, for both `L[i, i]` and `L[i, j]`,  wouldn't be a bad idea? As of now, the diagonal entries come out as expected because of Abs, but the off-diagonals are still full of unevaluated products of complex numbers (when A is Hermitian).

Reply:
It looks like there is the same issue with the products `L[i, k]*L[j, k].conjugate()`. (The treatment of complex numbers should be improved.) I think `expand_mul` should be applied for the present.

Reply:
Perhaps a `ComplexNumber` class should be added to numbers.py. Its instances would consist of a pair of instances of Number (or its subclasses).

Reply:
Okay, I'll add `expand_mul` to get complex numbers to multiply. It is annoying that we have to do this. When I mentioned this in Gitter chat, Aaron pointed out #12192 which attempted ComplexFloat class, and was never finished.

Reply:
It seems that ComplexFloat class was intended as a SymPy version of Python's complex class. Its instances could also be in the ComplexNumber class that I had in mind (represented by a couple of Floats). That should not be hard to implement, but possibly a large number of tests should be revised.